app/controllers/api/v1/questionnaires_controller.rb

```diff
@@ -0,0 +1,169 @@
+    else
+      questionnaire_private = params[:questionnaire][:private] == 'true'
+      display_type = params[:questionnaire][:type].split('Questionnaire')[0]
+      begin
+        @questionnaire = Object.const_get(params[:questionnaire][:type]).new if Questionnaire::QUESTIONNAIRE_TYPES.include? params[:questionnaire][:type]
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+      begin
+        @questionnaire.private = questionnaire_private
+        @questionnaire.name = params[:questionnaire][:name]
+        @questionnaire.instructor_id = 6 # session[:user].id
+        @questionnaire.min_question_score = params[:questionnaire][:min_question_score]
+        @questionnaire.max_question_score = params[:questionnaire][:max_question_score]
+        @questionnaire.type = params[:questionnaire][:type]
+        if %w[AuthorFeedback CourseSurvey TeammateReview GlobalSurvey AssignmentSurvey BookmarkRating].include?(display_type)
+          display_type = (display_type.split(/(?=[A-Z])/)).join('%')
+        end
+        @questionnaire.display_type = display_type
+        @questionnaire.instruction_loc = Questionnaire::DEFAULT_QUESTIONNAIRE_URL
+        @questionnaire.save
+        # Create node
+        tree_folder = TreeFolder.where(['name like ?', @questionnaire.display_type]).first
+        parent = FolderNode.find_by(node_object_id: tree_folder.id)
+        QuestionnaireNode.create(parent_id: parent.id, node_object_id: @questionnaire.id, type: 'QuestionnaireNode')
+        render json: @questionnaire, status: :created
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+     
+    end
+  end
+
+  # Remove a given questionnaire
+  # DELETE on /questionnaires/:id
+  def destroy
+    begin
+      @questionnaire = Questionnaire.find(params[:id])
+    rescue
+      render json: $ERROR_INFO
+    end
+    if @questionnaire
+      begin
+        name = @questionnaire.name
+        questions = @questionnaire.questions
+        questions.each do |question|
+          advices = question.question_advices
+          advices.each(&:delete)
+          question.delete
+        end
+        questionnaire_node = @questionnaire.questionnaire_node
+        questionnaire_node.delete if !questionnaire_node.nil?
+        @questionnaire.delete
+        render json: "The questionnaire \"#{name}\" has been successfully deleted."
+      rescue StandardError => e
+        render json: e.message
+      end
+    end
+  end
+
+  #Update a given questionnaire
+  # PUT on /questionnaires/:id
+  def update
+    # If 'Add' or 'Edit/View advice' is clicked, redirect appropriately
+      begin
+        # Save questionnaire information
+        @questionnaire = Questionnaire.find(params[:id])
+        @questionnaire.update_attributes(questionnaire_params)
+        # Save all questions
+        unless params[:question].nil?
+          params[:question].each_pair do |k, v|
```

Comment:
need to handle it better, may be as part of the user interface similar to delete

Reply:
Commit 285c94a addresses this review.